default/python/AI/SaveGameManager/__init__.py

```diff
@@ -0,0 +1,47 @@
+"""This package provides json-based encoding and decoding for use in FreeOrion AI savegames.
+
+The encoding is json-based with custom prefixes to support some objects
+and dictionary keys of types which are not supported in standard json.
+
+The decoder will only load trusted classes as defined in _definitions.py,
+if an unknown/untrusted object is encountered, it will raise a InvalidSaveGameException.
+
+When class instances are loaded, the __setstate__ method will be invoked if available or
+the __dict__ content will be set directly. It is the responsibility of each trusted class
+to define a __setstate__ method to verify and possibly sanitize the passed data.
+The __setstate__ method is expected to raise an Exception if invalid data was passed.
+
+In contrast to standard library json module, the type of dict keys (e.g. int) is preserved:
+    import json
+    json_string = json.dumps({1: 2})  ==> '{"1": 2}'
+    json.loads(json_string) ==> {"1": 2}
+
+    import SaveGameManager  # load package
+    json_string = SaveGameManager.encode({1: 2})  ==> '{"__INT__1": 2'}'
+    SaveGameManager.decode(json_string)  ==> {1: 2}
+
+
+Public functions:
+    encode  -  encode a python object to an unambigious string representation
+    decode  -  decode and interpret a json string to retreive python objects
+    build_savegame_string  - encode the AIstate and return the zlib-compressed result
```

Comment:
It could be helpful to support uncompressed save state strings, to allow inspection of save files when saved in uncompressed XML format. This could be toggled by a flag included in each save data request, or as part of each save state request from the server.

The AI save state is already stored within the compressed portion of the XML format saves, so compressing it again won't help much for file size, though it should reduce the amount of data sent over network connection, so doing so is perhaps useful for that purpose.